programs/ziti-edge-tunnel/netif_driver/windows/tun.c

```diff
@@ -0,0 +1,262 @@
+    X(WintunReleaseReceivePacket, WINTUN_RELEASE_RECEIVE_PACKET_FUNC) ||
+    X(WintunAllocateSendPacket, WINTUN_ALLOCATE_SEND_PACKET_FUNC) ||
+    X(WintunSendPacket, WINTUN_SEND_PACKET_FUNC))
+#undef X
+    {
+        DWORD LastError = GetLastError();
+        FreeLibrary(Wintun);
+        SetLastError(LastError);
+        Wintun = NULL;
+    }
+
+    WINTUN = Wintun;
+}
+
+netif_driver tun_open(struct uv_loop_s *loop, uint32_t tun_ip, uint32_t dns_ip, const char *cidr, char *error, size_t error_len) {
+    if (error != NULL) {
+        memset(error, 0, error_len * sizeof(char));
+    }
+
+    uv_once(&wintunInit, InitializeWintun);
+    if (WINTUN == NULL) {
+        strcpy_s(error, error_len, "Failed to load wintun.dll");
+        return NULL;
+    }
+
+    struct netif_handle_s *tun = calloc(1, sizeof(struct netif_handle_s));
+    if (tun == NULL) {
+        if (error != NULL) {
+            snprintf(error, error_len, "failed to allocate tun");
+        }
+        return NULL;
+    }
+
+    GUID adapterGuid;
+    IIDFromString(L"2cbfd72d-370c-43b0-b0cd-c8f092a7e134", &adapterGuid);
+    tun->adapter = WintunOpenAdapter(L"Ziti", L"tun0");
+    if (!tun->adapter) {
+        tun->adapter = WintunCreateAdapter(L"Ziti", L"tun0", &adapterGuid, NULL);
+        if (!tun->adapter) {
+            DWORD err = GetLastError();
+            snprintf(error, error_len, "Failed to create adapter: %d", err);
+            return NULL;
+        }
+    }
+
+    DWORD Version = WintunGetRunningDriverVersion();
+    ZITI_LOG(INFO, "Wintun v%u.%u loaded", (Version >> 16) & 0xff, (Version >> 0) & 0xff);
+
+    MIB_UNICASTIPADDRESS_ROW AddressRow;
+    InitializeUnicastIpAddressEntry(&AddressRow);
+    WintunGetAdapterLUID(tun->adapter, &AddressRow.InterfaceLuid);
+    AddressRow.Address.Ipv4.sin_family = AF_INET;
+    AddressRow.Address.Ipv4.sin_addr.S_un.S_addr = tun_ip;
+    AddressRow.OnLinkPrefixLength = 16;
+    DWORD err = CreateUnicastIpAddressEntry(&AddressRow);
+    if (err != ERROR_SUCCESS && err != ERROR_OBJECT_ALREADY_EXISTS)
+    {
+        ZITI_LOG(ERROR, "Failed to set IP address: %d", err);
+        tun_close(tun);
+        return NULL;
+    }
+
+    tun->session = WintunStartSession(tun->adapter, WINTUN_MAX_RING_CAPACITY);
+    if (!tun->session) {
+        err = GetLastError();
+        snprintf(error, error_len, "Failed to start session: %d", err);
+        return NULL;
+    }
+
+    struct netif_driver_s *driver = calloc(1, sizeof(struct netif_driver_s));
+    if (driver == NULL) {
+        if (error != NULL) {
+            snprintf(error, error_len, "failed to allocate netif_device_s");
+            tun_close(tun);
+        }
+        return NULL;
+    }
+    ZITI_LOG(DEBUG, "tun=%p, adapter=%p, session=%p", tun, tun->adapter, tun->session);
+
+    driver->handle       = tun;
+    driver->setup        = tun_setup_read;
+    driver->write        = tun_write;
+    driver->add_route    = NULL; // TODO tun_add_route;
```

Comment:
wireguard-windows has a bunch of helper functions that you could use to discover the proper C API functions if you want to use that for inspiration. specifically functions like https://git.zx2c4.com/wireguard-windows/tree/tunnel/winipcfg/luid.go#n210 called from the go version at https://github.com/openziti/desktop-edge-win/blob/a02a50cf6437a956329acd409e5fee7923461dd2/service/ziti-tunnel/service/state.go#L430

Reply:
there's a DeleteRoute in luid.go too that is used currently by the desktop-edge-win